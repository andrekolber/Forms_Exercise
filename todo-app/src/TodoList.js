import React, { useState } from 'react';
import NewTodoForm from './NewTodoForm';
import { v4 as uuid } from 'uuid';
import Todo from './Todo';
import './TodoList.css';

const TodoList = () => {
	const INITIAL_STATE = [];
	const [ todos, setTodos ] = useState(INITIAL_STATE);

	const addTodo = (newTodo) => {
		setTodos((todos) => [ ...todos, { id: uuid(), ...newTodo } ]);
	};

	const removeTodo = (id) => {
		setTodos((todos) => todos.filter((todo) => todo.id !== id));
	};
	return (
		<div>
			<h3 className="h3">Todo List</h3>
			<NewTodoForm addTodo={addTodo} />
			{todos.map(({ id, todo }) => <Todo todo={todo} id={id} key={id} removeTodo={removeTodo} />)}
		</div>
	);
};

export default TodoList;
