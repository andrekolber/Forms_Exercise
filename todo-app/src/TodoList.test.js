import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import TodoList from './TodoList';

const addTodo = (todoList, todo = 'Walk the Dog') => {
	const todoInput = todoList.getByLabelText('New Todo:');

	fireEvent.change(todoInput, { target: { value: todo } });

	const button = todoList.getByText('Add');
	fireEvent.click(button);
};

it('renders without crashing', () => {
	render(<TodoList />);
});

it('matches snapshot', () => {
	const { asFragment } = render(<TodoList />);
	expect(asFragment()).toMatchSnapshot();
});

it('can add a new todo', () => {
	const todoList = render(<TodoList />);

	// no todos yet
	expect(todoList.queryByText('X')).not.toBeInTheDocument();

	addTodo(todoList);

	// expect todo to exist
	const removeBtn = todoList.queryByText('X');
	const todo = todoList.getByRole('todo');
	expect(removeBtn).toBeInTheDocument();
	expect(todo).toBeInTheDocument();

	// expect form to be empty
	expect(todoList.getAllByDisplayValue('')).toHaveLength(1);
});

it('can remove a todo', () => {
	const todoList = render(<TodoList />);
	addTodo(todoList);

	const removeBtn = todoList.getByText('X');

	// click the remove button and the todo should dissapear
	fireEvent.click(removeBtn);
	expect(removeBtn).not.toBeInTheDocument();
});
