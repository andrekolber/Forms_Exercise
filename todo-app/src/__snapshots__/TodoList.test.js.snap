// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <h3
      class="h3"
    >
      Todo List
    </h3>
    <form
      class="NewTodoForm"
    >
      <label
        for="todo"
      >
        New Todo: 
      </label>
      <input
        id="todo"
        name="todo"
        placeholder="Add New Todo"
        type="text"
        value=""
      />
      <button
        class="NewTodoForm-Btn"
      >
        Add
      </button>
    </form>
  </div>
</DocumentFragment>
`;
